telegram/ext/filters.py

```diff
@@ -892,38 +893,153 @@ class user(BaseFilter):
+            return set(user_id)
+
+        @staticmethod
+        def _parse_username(username):
             if username is None:
-                self.usernames = username
-            elif isinstance(username, string_types):
-                self.usernames = [username.replace('@', '')]
-            else:
-                self.usernames = [user.replace('@', '') for user in username]
+                return set()
+            if isinstance(username, str):
+                return {username.replace('@', '')}
+            return {user.replace('@', '') for user in username}
+
+        def _set_user_ids(self, user_id):
+            with self.__lock:
+                if user_id and self._usernames:
+                    raise RuntimeError("Can't set user_id in conjunction with (already set) "
+                                       "usernames.")
+                self._user_ids = self._parse_user_id(user_id)
+
+        def _set_usernames(self, username):
+            with self.__lock:
+                if username and self._user_ids:
+                    raise RuntimeError("Can't set username in conjunction with (already set) "
+                                       "user_ids.")
+                self._usernames = self._parse_username(username)
+
+        @property
+        def user_ids(self):
+            with self.__lock:
+                return frozenset(self._user_ids)
+
+        @user_ids.setter
+        def user_ids(self, user_id):
+            self._set_user_ids(user_id)
+
+        @property
+        def usernames(self):
+            with self.__lock:
+                return frozenset(self._usernames)
+
+        @usernames.setter
+        def usernames(self, username):
+            self._set_usernames(username)
+
+        def add_users(self, user_id=None, username=None):
+            """
+            Adds a user(s) to the allowed users.
+
+            Args:
+                user_id(:obj:`int` | List[:obj:`int`], optional): Which user ID(s) to allow
+                    through.
+                username(:obj:`str` | List[:obj:`str`], optional): Which username(s) to allow
+                    through. If username starts with '@' symbol, it will be ignored.
+            """
+            with self.__lock:
+                if user_id and self._usernames:
+                    raise RuntimeError("Can't set user_id in conjunction with (already set) "
+                                       "usernames.")
+                if username and self._user_ids:
+                    raise RuntimeError("Can't set username in conjunction with (already set) "
+                                       "user_ids.")
+
+                user_id = self._parse_user_id(user_id)
+                username = self._parse_username(username)
+
+                for uid in user_id:
+                    self._user_ids.add(uid)
+                for name in username:
+                    self._usernames.add(name)
+
+        def remove_users(self, user_id=None, username=None):
```

Comment:
any reason not to make it a dedicated method for remove_usernames and dedicated method for remove_user_ids?

Reply:
Not really. Changed.